docker/nginx.js

```diff
@@ -12,52 +21,153 @@ function proxyJolokiaAgent(req) {
-    req.subrequest(`/authorization/namespaces/${namespace}/localsubjectaccessreviews`,
-      {
-        method: 'POST',
-        body: JSON.stringify({
-          kind: 'LocalSubjectAccessReview',
-          apiVersion: 'authorization.openshift.io/v1',
-          namespace: namespace,
-          verb: 'update',
-          resource: 'pods',
-          name: pod,
-        }),
-      },
-      function (res) {
-        res.status === 201
-          ? checkAuthorization(JSON.parse(res.responseBody))
-          : response(res);
-      });
+  function reject(status, message) {
+    return Promise.reject({
+      status: status,
+      responseBody: message,
+      headersOut: {
+        'Content-Type': 'application/json',
+      }
+    });
   }
 
-  function checkAuthorization(sar) {
-    if (sar.allowed) {
-      getPodIP();
-    } else {
-      req.return(403, sar.reason);
-    }
+  function selfLocalSubjectAccessReview(verb) {
+    // Work-around same-location sub-requests caching issue
+    return req.subrequest(`/authorization${verb === 'get' ? '2' : ''}/namespaces/${namespace}/localsubjectaccessreviews`, {
+      method: 'POST',
+      body: JSON.stringify({
+        kind: 'LocalSubjectAccessReview',
+        apiVersion: 'authorization.openshift.io/v1',
+        namespace: namespace,
+        verb: verb,
+        resource: 'pods',
+        name: pod,
+      }),
+    });
   }
 
   function getPodIP() {
-    req.subrequest(`/podIP/${namespace}/${pod}`, { method: 'GET' },
-      function (res) {
-        res.status === 200
-          ? callJolokiaAgent(JSON.parse(res.responseBody).status.podIP)
-          : response(res);
-      });
+    return req.subrequest(`/podIP/${namespace}/${pod}`, { method: 'GET' }).then(res => {
+      if (res.status !== 200) {
+        return Promise.reject(res);
+      }
+      return JSON.parse(res.responseBody).status.podIP;
+    });
+  }
+
+  // This is usually called once upon the front-end loads, still we may want to cache it
+  function listMBeans(podIP) {
+    return req.subrequest(`/proxy/${protocol}:${podIP}:${port}/${path}`, { method: 'POST', body: JSON.stringify({ type: 'list' }) }).then(res => {
+      if (res.status !== 200) {
+        return Promise.reject(res);
+      }
+      return JSON.parse(res.responseBody).value;
+    });
   }
 
-  function callJolokiaAgent(podIP) {
-    req.subrequest(`/proxy/${protocol}:${podIP}:${port}/${path}`, { method: req.method, body: req.requestBody }, response);
+  function callJolokiaAgent(podIP, request) {
+    return req.subrequest(`/proxy/${protocol}:${podIP}:${port}/${path}`, { method: req.method, body: request });
   }
 
-  selfLocalSubjectAccessReview();
+  return selfLocalSubjectAccessReview('update')
+    .then(function (res) {
+      if (res.status !== 201) {
+        return Promise.reject(res);
+      }
+      var sar = JSON.parse(res.responseBody);
+      if (sar.allowed) {
+        // map the 'update' verb to the 'admin' role
+        return 'admin';
```

Comment:
So it might have been described somewhere earlier already, but do we support only the fixed set of roles which is admin and viewer now, and will it remain so in the future as well?  It's not a big deal as I'm sure we've captured the customer's requirements being boiled down to the two roles; I just wanted to understand the impl correctly.

Reply:
The document _Fuse Console on OpenShift RBAC Support_ appears to explain it already:
> The idea of the proposed solution is to evolve step 3, from a unary RBAC mechanism, with only one assimilated admin, into a binary one, supporting two roles, viewer and admin, inferred from the permissions the user is granted on the Fuse application pod.

Reply:
I've also added some documentation in eba473e2bd1e511bc8f180d1921917cbc4268d09.